pkg/trustedresources/verify.go

```
@@ -100,18 +89,99 @@ func VerifyPipeline(ctx context.Context, pipelineObj v1beta1.PipelineObject, k8s
+// 3. To pass one policy, the resource can pass any public keys in the policy.
+func verifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, signature []byte, source string, policies []*v1alpha1.VerificationPolicy) error {
+	verifiers := verifiersFromConfigMap(ctx, k8s)
+	if len(verifiers) != 0 {
+		for _, verifier := range verifiers {
+			if err := verifyInterface(resource, verifier, signature); err == nil {
+				return nil
+			}
+		}
+		return fmt.Errorf("resource %s in namespace %s fails verification", resource.GetName(), resource.GetNamespace())
+	}
+
+	if len(policies) == 0 {
+		return errorEmptyVerificationConfig
+	}
+
+	matchedPolicies := matchedPolicies(policies, source)
+	if len(matchedPolicies) == 0 {
+		return fmt.Errorf("no matching policies are found for resource: %s against source: %s. %w", resource.GetName(), source, errorNoMatchedPolicies)
+	}
+
+	verifiedPolicies, err := verifiedPolicies(ctx, resource, k8s, signature, matchedPolicies)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to verify the resource with matched policies: %w", err)
+	}
+
+	// make sure all matched policies pass verification
+	if verifiedPolicies == len(matchedPolicies) {
+		return nil
 	}
+	return fmt.Errorf("resource %s in namespace %s fails verification %w", resource.GetName(), resource.GetNamespace(), ErrorResourceVerificationFailed)
+
+}
 
-	for _, verifier := range verifiers {
-		if err := VerifyInterface(pipeline, verifier, signature); err == nil {
-			return nil
+// matchedPolicies return all matched VerificationPolicy by checking if the Pattern is matching source.
+func matchedPolicies(policies []*v1alpha1.VerificationPolicy, source string) []*v1alpha1.VerificationPolicy {
+	matchedPolicies := []*v1alpha1.VerificationPolicy{}
+	for _, vp := range policies {
+		for _, r := range vp.Spec.Resources {
+			matching, _ := regexp.MatchString(r.Pattern, source)
```

there's an error being swallowed here